File: 백준/Silver/1629. 곱셈/곱셈.py
import sys

# 분할정복
def solve(a, b, c):
    if b == 1:
        return a % c
 
    if b % 2 == 0:
        tmp = solve(a, b/2, c)
        return (tmp * tmp) % c
 
    else:
        tmp = solve(a, (b-1)/2, c)
        return (tmp * tmp * a) % c

a,b,c = map(int,sys.stdin.readline().split()) 
print(solve(a,b,c))

File: 백준/Silver/1629. 곱셈/곱셈.cc
#include <iostream>
using namespace std;

long long solve(long long a, long long b, long long c) {
    if (b == 1) return a % c;

    if (b % 2 == 0) {
        long long tmp = solve(a, b / 2, c);
        return (tmp * tmp) % c;
    } else {
        long long tmp = solve(a, (b - 1) / 2, c);
        return (tmp * tmp % c * a) % c;
    }
}

int main() {
    long long a, b, c;
    cin >> a >> b >> c;
    cout << solve(a, b, c) << endl;
    return 0;
}
